ipam/monitor/awsvpc.go

```diff
@@ -0,0 +1,265 @@
+	common.Log.Debugf("HandleUpdate: old(%q) new(%q)", old, new)
+
+	// It might make sense to do removal first and then add entries
+	// because of the 50 routes limit. However, in such case a container might
+	// not be reachable for short period of time which we we would like to
+	// avoid.
+
+	// Add new entries
+	for _, cidr := range newCIDRs {
+		cidrStr := cidr.String()
+		common.Log.Debugf("Creating %s route to %s", cidrStr, mon.instanceID)
+		_, err := mon.createVPCRoute(cidrStr)
+		// TODO(mp) check for 50 routes limit
+		// TODO(mp) maybe check for auth related errors
+		if err != nil {
+			return fmt.Errorf("createVPCRoutes failed: %s", err)
+		}
+		err = mon.createHostRoute(cidrStr)
+		if err != nil {
+			return fmt.Errorf("createHostRoute failed: %s", err)
+		}
+	}
+
+	// Remove obsolete entries
+	for _, cidr := range oldCIDRs {
+		cidrStr := cidr.String()
+		common.Log.Debugf("Removing %s route", cidrStr)
+		_, err := mon.deleteVPCRoute(cidrStr)
+		if err != nil {
+			return fmt.Errorf("deleteVPCRoute failed: %s", err)
+		}
+		err = mon.deleteHostRoute(cidrStr)
+		if err != nil {
+			return fmt.Errorf("deleteHostRoute failed: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func (mon *AWSVPCMonitor) String() string {
+	return "awsvpc"
+}
+
+func (mon *AWSVPCMonitor) createVPCRoute(cidr string) (*ec2.CreateRouteOutput, error) {
+	route := &ec2.CreateRouteInput{
+		RouteTableId:         &mon.routeTableID,
+		InstanceId:           &mon.instanceID,
+		DestinationCidrBlock: &cidr,
+	}
+	return mon.ec2.CreateRoute(route)
+}
+
+func (mon *AWSVPCMonitor) createHostRoute(cidr string) error {
+	dst, err := parseCIDR(cidr)
+	if err != nil {
+		return err
+	}
+	route := &netlink.Route{
+		LinkIndex: mon.linkIndex,
+		Dst:       dst,
+		Scope:     netlink.SCOPE_LINK,
+	}
+	return netlink.RouteAdd(route)
+}
+
+func (mon *AWSVPCMonitor) deleteVPCRoute(cidr string) (*ec2.DeleteRouteOutput, error) {
+	route := &ec2.DeleteRouteInput{
+		RouteTableId:         &mon.routeTableID,
+		DestinationCidrBlock: &cidr,
+	}
+	return mon.ec2.DeleteRoute(route)
+}
+
+func (mon *AWSVPCMonitor) deleteHostRoute(cidr string) error {
+	dst, err := parseCIDR(cidr)
+	if err != nil {
+		return err
+	}
+	route := &netlink.Route{
+		LinkIndex: mon.linkIndex,
+		Dst:       dst,
+		Scope:     netlink.SCOPE_LINK,
+	}
+	return netlink.RouteDel(route)
+}
+
+func (mon *AWSVPCMonitor) detectRouteTableID() (*string, error) {
+	instancesParams := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(mon.instanceID)},
+	}
+	instancesResp, err := mon.ec2.DescribeInstances(instancesParams)
+	if err != nil {
+		return nil, fmt.Errorf("DescribeInstances failed: %s", err)
+	}
+	if len(instancesResp.Reservations) == 0 ||
+		len(instancesResp.Reservations[0].Instances) == 0 {
+		return nil, fmt.Errorf(
+			"cannot find %s instance within reservations", mon.instanceID)
+	}
+	vpcID := instancesResp.Reservations[0].Instances[0].VpcId
```

Comment:
AWS VPC Route Table ID detection. Perhaps, I should comment this method properly.